source/common/stream_info/stream_info_impl.h

```diff
@@ -12,13 +12,26 @@
 namespace Envoy {
 namespace StreamInfo {
 
+// Mutable filter state wrapping metadata object. State within this
+// metadata should still be namespaced per usual reverse dns naming
```

Comment:
What's the tradeoff here between having the namespacing at the FilterState level and just embedding a Struct here? Shouldn't it be possible to still preserve the existing semantics this way?

Reply:
I'm less concerned about the churn to existing code as with getting the right long term abstraction in place. I do realize though that it's a bit unfair to make this statement without skin in the game, as I'm not doing the refactoring work. Would be interested in hearing from other folks like this, such as @mattklein123.

Reply:
Yeah tough call. What about one of these two options:
1) Mark `dynamicMetadata()` as deprecated and subject to removal in a future version. No new code should use it.
2) More aggressively, change the name to `deprecatedDynamicMetadata()` to really make sure people understand they shouldn't be using it? This would require some find/replace across the code and some trivial breakage to third party filters, but with some comments on the deprecated functions on how people should migrate away I think this would be a good thing? We could also open issues on all the existing call sites within the repo to be cleaned up as tech debt?

Reply:
That sounds good by me. That said, I would still propose to keep the well known FilterState called envoy.streaminfo.dynamicMetadata around. Allows people to pile stuff into envoy.streaminfo.dynamicmetadata filter State, where they still have to use soem well defined namespaces to put things in. At the same time, it allows consumers to inspect whats been stored in this dynamic filter state (map semantics), without needing to add all those interfaces to FilterState. 

From that point of view, would it make sense to rename this to `wellKnownDynamicFilterState()` ? Its still the same find/replace operation with no tech debt

Reply:
That sounds reasonable to me, but we should have liberal comments on when/why to use one or the other. @htuch WDYT?

Reply:
@mattklein123 @rshriram I'm still uncomfortable with this. I don't see how this achieves anything that can't be achieved with placing dynamic metadata directly under the relevant top-level namespace in filter state. We can support iteration on filter state as well if necessary.

Reply:
The question is should we sunset this under a deprecated namespace as you suggest or permanently reserve it? I worry about this being an anti-pattern going forward, where the grab bag is (mis)used by different filters. For example, imagine we have two Envoy filter vendors Acme and Emca. They have both have a set of 5 filters that use the grab bag for something completely different, e.g. Acme thinks this is a good place for logging data, Emca plans to use it to store internal filer state for Emca. These filters will no longer compose into a single chain.

Reply:
OK, I just chatted with @mattklein123 offline. I'm happy with a combination of https://github.com/envoyproxy/envoy/pull/4873#discussion_r228770364, adding a well known `envoy.streaminfo.dynamicmetadata` for those who really want it and a personal request to provide some well-throught out guidance on the pros and cons of using a global grab big vs. namespaced data. I'm happy to sign-off if we have all three.

Reply:
I just talked to @htuch offline and I agree we should deprecate it, and file follow up tech debt issues to fix both access logging and Lua to specify both a namespace and item when access the struct. 

Beyond this, if people want to use a grab bag in a well known namespace that is up to them.

Is that fair?

Reply:
What to do about all the users of metadata match thats is a common user of the metadata struct obtained from Dynamic metadata? Provide it with full access to the internal state in the map? Even then, its going to be full of exception catch code if the FilterState doesn't match the Struct object type.

Reply:
I think for this case we can avoid exception catch by allowing code to test to see if the map contains a struct type.

Reply:
Yeah, but I am not sure Randy and co would be happy with that given that they wanted super strict semantics on how data can be accessed across filters